presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchangeForSingleWorkerExecution.java

```diff
@@ -0,0 +1,160 @@
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.session = requireNonNull(session, "session is null");
+        }
+
+        public boolean isPlanChanged()
+        {
+            return planChanged;
+        }
+
+        @Override
+        public PlanNode visitTableScan(TableScanNode node, RewriteContext<Void> context)
+        {
+            PlanNode plan = pushPredicateIntoTableScan(node, new ConstantExpression(true, BOOLEAN), true, session, idAllocator, metadata);
+            // Presto Java and Presto Native use different hash functions for partitioning
+            // An additional exchange makes sure the data flows through a native worker in case it need to be partitioned for downstream processing
+            if (containsSystemTableScan(plan)) {
+                plan = gatheringExchange(idAllocator.getNextId(), REMOTE_STREAMING, plan);
+            }
+            return plan;
+        }
+
+        @Override
+        public PlanNode visitExplainAnalyze(ExplainAnalyzeNode node, RewriteContext<Void> context)
+        {
+            return addGatherExchange(node);
+        }
+
+        @Override
+        public PlanNode visitTableFinish(TableFinishNode node, RewriteContext<Void> context)
+        {
+            PlanNode child = node.getSource();
+
+            ExchangeNode gather;
+            // In case the input is a union (see PushTableWriteThroughUnion), don't add another exchange. Instead, convert it to gather exchange
+            if (child instanceof ExchangeNode) {
+                ExchangeNode exchangeNode = (ExchangeNode) child;
+                gather = new ExchangeNode(
+                        exchangeNode.getSourceLocation(),
+                        idAllocator.getNextId(),
+                        GATHER,
+                        REMOTE_STREAMING,
+                        new PartitioningScheme(Partitioning.create(SINGLE_DISTRIBUTION, ImmutableList.of()), exchangeNode.getOutputVariables()),
+                        exchangeNode.getSources(),
+                        exchangeNode.getInputs(),
+                        true,
+                        Optional.empty());
+            }
+            else {
+                gather = ensureSourceOrderingGatheringExchange(idAllocator.getNextId(), REMOTE_STREAMING, child);
+            }
+            planChanged = true;
+            return ChildReplacer.replaceChildren(node, ImmutableList.of(gather));
+        }
+
+        @Override
+        public PlanNode visitStatisticsWriterNode(StatisticsWriterNode node, RewriteContext<Void> context)
+        {
+            return addGatherExchange(node);
+        }
+
+        private PlanNode addGatherExchange(InternalPlanNode node)
+        {
+            PlanNode child = node.getSources().get(0);
+
+            // if the child is already a gathering exchange
+            if ((child instanceof ExchangeNode) && ((ExchangeNode) child).getType() == ExchangeNode.Type.GATHER) {
```

Comment:
nit: Maybe also check if child is an exchange and replace it with `GATHER`. Then you can use it in `visitTableFinish` as well.